modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PlannerHelper.java

```diff
@@ -343,4 +356,145 @@ static boolean hasSubQuery(SqlNode node) {
             return super.visit(call);
         }
     }
+
+
+    /**
+     * Tries to optimize a query that looks like {@code SELECT count(*)}.
+     *
+     * @param planner Planner.
+     * @param txContext Transactional context.
+     * @param node Query node.
+     * @return Plan node with list of aliases, if the optimization is applicable.
+     */
+    public static @Nullable Pair<IgniteRel, List<String>> tryOptimizeSelectCount(
+            IgnitePlanner planner,
+            @Nullable QueryTransactionContext txContext,
+            SqlNode node
+    ) {
+        if (txContext != null && txContext.explicitTx() != null) {
+            return null;
+        }
+
+        if (!(node instanceof SqlSelect)) {
+            return null;
+        }
+
+        SqlSelect select = (SqlSelect) node;
+
+        if (select.getGroup() != null
+                || select.getWhere() != null
+                || select.getHaving() != null
+                || select.getQualify() != null
+                || !select.getWindowList().isEmpty()
+                || select.getOrderList() != null && !select.getOrderList().isEmpty()
```

Comment:
I see you've added orderList here. What is wrong with `ORDER BY`? I mean, at first glance ordering doesn't affect result of COUNT. Do I miss something?

Reply:
Thanks. Fixed.